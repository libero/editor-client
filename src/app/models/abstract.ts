import { EditorState } from 'prosemirror-state';
import { DOMParser as ProseMirrorDOMParser } from 'prosemirror-model';
import { gapCursor } from 'prosemirror-gapcursor';

import { makeSchemaFromConfig } from 'app/models/utils';
import * as abstractConfig from 'app/models/config/abstract.config';
import { buildInputRules } from 'app/models/plugins/input-rules';
import { PlaceholderPlugin } from 'app/models/plugins/placeholder.plugin';
import { SelectionPlugin } from 'app/models/plugins/selection.plugins';

export function createAbstractState(content: Element): EditorState {
  const schema = makeSchemaFromConfig(abstractConfig.topNode, abstractConfig.nodes, abstractConfig.marks);
  const xmlContentDocument = document.implementation.createDocument('', '', null);

  if (content) {
    xmlContentDocument.appendChild(content);
  }

  return EditorState.create({
    doc: ProseMirrorDOMParser.fromSchema(schema).parse(xmlContentDocument),
    schema,
    plugins: [buildInputRules(), SelectionPlugin, gapCursor(), PlaceholderPlugin('Enter abstract')]
  });
}

export function createImpactStatementState(content: Element): EditorState {
  const schema = makeSchemaFromConfig(abstractConfig.topNode, abstractConfig.nodes, abstractConfig.marks);
  const xmlContentDocument = document.implementation.createDocument('', '', null);

  if (content) {
    xmlContentDocument.appendChild(content);
  }

  return EditorState.create({
    doc: ProseMirrorDOMParser.fromSchema(schema).parse(xmlContentDocument),
    schema,
    plugins: [buildInputRules(), SelectionPlugin, gapCursor(), PlaceholderPlugin('Enter impact statement')]
  });
}
