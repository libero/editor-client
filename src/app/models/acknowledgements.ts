import { DOMParser as ProseMirrorDOMParser } from 'prosemirror-model';
import { EditorState } from 'prosemirror-state';
import { Step } from 'prosemirror-transform';
import { gapCursor } from 'prosemirror-gapcursor';

import * as acknowledgementsConfig from './config/acknowledgements.config';
import { buildInputRules } from './plugins/input-rules';
import { makeSchemaFromConfig } from './utils';
import { PlaceholderPlugin } from './plugins/placeholder.plugin';
import { SelectionPlugin } from './plugins/selection.plugins';

export function createAcknowledgementsState(content?: Element, changeSteps?: [Step]): EditorState {
  if (content) {
    const ackTitle = content.querySelector('title');
    if (ackTitle) {
      content.removeChild(ackTitle);
    }
  }

  const schema = makeSchemaFromConfig(
    acknowledgementsConfig.topNode,
    acknowledgementsConfig.nodes,
    acknowledgementsConfig.marks
  );
  const xmlContentDocument = document.implementation.createDocument('', '', null);

  if (content) {
    xmlContentDocument.appendChild(content);
  }

  return EditorState.create({
    doc: ProseMirrorDOMParser.fromSchema(schema).parse(xmlContentDocument),
    schema,
    plugins: [buildInputRules(), SelectionPlugin, gapCursor(), PlaceholderPlugin('Enter acknowledgements')]
  });
}
