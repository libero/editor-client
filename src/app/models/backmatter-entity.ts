import { v4 as uuidv4 } from 'uuid';

import { JSONObject } from '../types/utility.types';

export abstract class BackmatterEntity {
  constructor() {
    this._id = uuidv4();
  }

  get id(): string {
    return this._id;
  }

  protected _id: string;

  protected abstract fromXML(xmlNode: Element): void;
  protected abstract fromJSON(json: JSONObject): void;
  protected abstract createBlank(): void;

  protected createEntity(data?: Element | JSONObject): void {
    if (!data) {
      this.createBlank();
    } else if (data instanceof Element) {
      this.fromXML(data);
    } else {
      this.fromJSON(data);
    }
  }
}
