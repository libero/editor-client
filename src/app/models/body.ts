import { DOMParser as ProseMirrorDOMParser } from 'prosemirror-model';
import { EditorState } from 'prosemirror-state';
import { gapCursor } from 'prosemirror-gapcursor';
import { dropCursor } from 'prosemirror-dropcursor';
import { keymap } from 'prosemirror-keymap';
import { baseKeymap } from 'prosemirror-commands';
import { set } from 'lodash';

import * as bodyConfig from './config/body.config';
import { buildInputRules } from './plugins/input-rules';
import { makeSchemaFromConfig } from 'app/models/utils';
import { PlaceholderPlugin } from 'app/models/plugins/placeholder.plugin';
import { createListKeymap } from 'app/utils/prosemirror/list.helpers';
import { SelectionPlugin } from 'app/models/plugins/selection.plugins';

export function createBodyState(content: Element, id: string): EditorState {
  const schema = makeSchemaFromConfig(bodyConfig.topNode, bodyConfig.nodes, bodyConfig.marks);
  const xmlContentDocument = document.implementation.createDocument('', '', null);

  if (content) {
    xmlContentDocument.appendChild(content);
  }

  set(xmlContentDocument, 'manuscriptPath', `/manuscripts/${id}`);

  return EditorState.create({
    doc: ProseMirrorDOMParser.fromSchema(schema).parse(xmlContentDocument),
    schema,
    plugins: [
      buildInputRules(),
      gapCursor(),
      SelectionPlugin,
      dropCursor({ color: '#0078CF', class: 'drop-cursor' }),
      keymap(createListKeymap(schema)),
      keymap(baseKeymap),
      PlaceholderPlugin('Enter main text')
    ]
  });
}
