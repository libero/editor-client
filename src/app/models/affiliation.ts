import { getTextContentFromPath } from 'app/models/utils';
import { BackmatterEntity } from 'app/models/backmatter-entity';
import { JSONObject } from 'app/types/utility.types';

export class Affiliation extends BackmatterEntity {
  label: string;
  institution: { name: string };
  address: { city: string };
  country: string;

  constructor(data?: JSONObject | Element) {
    super();
    this.createEntity(data);
  }

  public getDisplayName(): string {
    return [this.institution.name, this.address.city, this.country].filter(Boolean).join(', ');
  }

  public clone(): Affiliation {
    const affitiation = new Affiliation();
    affitiation._id = this._id;
    affitiation.label = this.label;
    affitiation.country = this.country;
    affitiation.institution = { name: this.institution.name };
    affitiation.address = { city: this.address.city };
    return affitiation;
  }

  protected fromXML(xml: Element): void {
    this._id = xml.getAttribute('id');
    this.label = getTextContentFromPath(xml, 'label');
    this.institution = {
      name: [
        getTextContentFromPath(xml, 'institution[content-type="dept"]'),
        getTextContentFromPath(xml, 'institution:not([content-type])')
      ]
        .filter(Boolean)
        .join(', ')
    };

    this.address = {
      city: getTextContentFromPath(xml, 'addr-line named-content[content-type="city"]')
    };
    this.country = getTextContentFromPath(xml, 'country');
  }

  protected fromJSON(json: JSONObject): void {
    this._id = json.id as string;
    this.label = json.label as string;
    this.country = json.country as string;
    this.institution = json.institution as { name: string };
    this.address = json.address as { city: string };
  }

  protected createBlank(): void {
    this.label = '';
    this.country = '';
    this.institution = { name: '' };
    this.address = { city: '' };
  }
}

export function createAffiliationsState(affiliationsXml: Element[]): Affiliation[] {
  return affiliationsXml.map((xml) => new Affiliation(xml));
}
