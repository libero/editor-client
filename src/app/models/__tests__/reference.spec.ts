import { createReferencesState, Reference, sortReferencesList } from 'app/models/reference';
import { set } from 'lodash';
import { WebReference } from 'app/models/reference-type';

jest.mock('uuid', () => ({
  v4: () => 'unique_id'
}));

describe('Reference model', () => {
  it('creates journal reference', () => {
    const el = document.createElement('p');
    el.innerHTML = givenReferenceXml(
      `
        <year iso-8601-date="2012">2012</year>
        <article-title>Assembly principles of <italic>Vibrio cholerae</italic> biofilms</article-title>
        <source>Science</source>
        <volume>337</volume>
        <fpage>236</fpage>
        <lpage>239</lpage>`,
      'journal'
    );

    el.setAttribute('id', 'id');
    expect(new Reference(el.querySelector('element-citation'))).toMatchSnapshot();
  });

  it('creates book reference', () => {
    const el = document.createElement('p');
    el.innerHTML = givenReferenceXml(
      `
        <year iso-8601-date="2010">2010</year>
        <source>Against Method</source>
        <edition>4th Edition</edition>
        <publisher-loc>London</publisher-loc>
        <publisher-name>Verso</publisher-name>
        <pub-id pub-id-type="isbn">978-1844674428</pub-id>`,
      'book'
    );

    el.setAttribute('id', 'id');
    expect(new Reference(el.querySelector('element-citation'))).toMatchSnapshot();
  });

  it('creates periodical reference', () => {
    const el = document.createElement('p');
    el.innerHTML = givenReferenceXml(
      `
        <string-date><month>September</month> <day>9</day>, <year iso-8601-date="1993-09-09">1993</year></string-date>
        <article-title>Obesity affects economic, social status</article-title>
        <source>The Washington Post</source>
        <fpage>1</fpage>
        <lpage>4</lpage>`,
      'periodical'
    );

    el.setAttribute('id', 'id');
    expect(new Reference(el.querySelector('element-citation'))).toMatchSnapshot();
  });

  it('creates report reference', () => {
    const el = document.createElement('p');
    el.innerHTML = givenReferenceXml(
      `
        <year iso-8601-date="2015">2015</year>
        <source>World Malaria Report 2015</source>
        <publisher-loc>Geneva</publisher-loc>
        <publisher-name>World Health Organization</publisher-name>
        <ext-link ext-link-type="uri" xlink:href="http://www.who.int/malaria/publications/world-malaria-report-2015/en/">http://www.who.int/malaria/publications/world-malaria-report-2015/en/</ext-link>
      `,
      'report'
    );

    el.setAttribute('id', 'id');
    expect(new Reference(el.querySelector('element-citation'))).toMatchSnapshot();
  });

  it('creates report reference', () => {
    const el = document.createElement('p');
    el.innerHTML = givenReferenceXml(
      `
        <year iso-8601-date="2015">2015</year>
        <source>World Malaria Report 2015</source>
        <publisher-loc>Geneva</publisher-loc>
        <publisher-name>World Health Organization</publisher-name>
        <ext-link ext-link-type="uri" xlink:href="http://www.who.int/malaria/publications/world-malaria-report-2015/en/">http://www.who.int/malaria/publications/world-malaria-report-2015/en/</ext-link>
      `,
      'report'
    );
    el.setAttribute('id', 'id');
    expect(new Reference(el.querySelector('element-citation'))).toMatchSnapshot();
  });

  it('creates data reference', () => {
    const el = document.createElement('p');
    el.innerHTML = givenReferenceXml(
      `
        <year iso-8601-date="2014">2014</year>
        <data-title>Mus musculus T-box 2 (Tbx2), mRNA</data-title>
        <source>NCBI Nucleotide</source>
        <pub-id pub-id-type="accession" assigning-authority="NCBI" xlink:href="http://www.ncbi.nlm.nih.gov/nuccore/120407038">NM_009324</pub-id>
        <version designator="NM_009324.2">NM_009324.2</version>`,
      'data'
    );

    el.setAttribute('id', 'id');
    expect(new Reference(el.querySelector('element-citation'))).toMatchSnapshot();
  });

  it('creates web reference', () => {
    const el = document.createElement('p');
    el.innerHTML = givenReferenceXml(
      `
        <year iso-8601-date="2019">2019</year>
        <article-title>James Watson had a chance to salvage his reputation on race. He made things worse</article-title>
        <source>The New York Times</source>
        <ext-link ext-link-type="uri" xlink:href="https://www.nytimes.com/2019/01/01/science/watson-dna-genetics-race.html">https://www.nytimes.com/2019/01/01/science/watson-dna-genetics-race.html</ext-link>
        <date-in-citation iso-8601-date="2019-01-01">January 1, 2019</date-in-citation>`,
      'web'
    );

    el.setAttribute('id', 'id');
    expect(new Reference(el.querySelector('element-citation'))).toMatchSnapshot();
  });

  it('creates preprint reference', () => {
    const el = document.createElement('p');
    el.innerHTML = givenReferenceXml(
      `
        <year iso-8601-date="2019">2019</year>
        <article-title>Gender and international diversity improves equity in peer review</article-title>
        <source>bioRxiv</source>
        <pub-id pub-id-type="doi">10.1101/400515</pub-id>`,
      'preprint'
    );

    el.setAttribute('id', 'id');
    expect(new Reference(el.querySelector('element-citation'))).toMatchSnapshot();
  });

  it('creates software reference', () => {
    const el = document.createElement('p');
    el.innerHTML = givenReferenceXml(
      `
        <year iso-8601-date="2014">2014</year>
        <data-title>R: A Language and Environment for Statistical Computing</data-title>
        <publisher-loc>Vienna, Austria</publisher-loc>
        <publisher-name>R Foundation for Statistical Computing</publisher-name>
        <ext-link ext-link-type="uri" xlink:href="http://www.R-project.org/">http://www.R-project.org/</ext-link>`,
      'software'
    );

    el.setAttribute('id', 'id');
    expect(new Reference(el.querySelector('element-citation'))).toMatchSnapshot();
  });

  it('creates conference reference', () => {
    const el = document.createElement('p');
    el.innerHTML = givenReferenceXml(
      `
        <year iso-8601-date="2014">2014</year>
        <article-title>Automated hypothesis generation based on mining scientific literature</article-title>
        <conf-name>Proceedings of the 20th ACM SIGKDD International Conference on Knowledge Discovery and Data Mining</conf-name>
        <fpage>1877</fpage>
        <lpage>1886</lpage>
        <pub-id pub-id-type="doi">10.1145/2623330.2623667</pub-id>`,
      'confproc'
    );

    el.setAttribute('id', 'id');
    expect(new Reference(el.querySelector('element-citation'))).toMatchSnapshot();
  });

  it('creates thesis reference', () => {
    const el = document.createElement('p');
    el.innerHTML = givenReferenceXml(
      `
        <year iso-8601-date="2014">2014</year>
        <article-title>Automated hypothesis generation based on mining scientific literature</article-title>
        <publisher-loc>Vienna, Austria</publisher-loc>
        <publisher-name>R Foundation for Statistical Computing</publisher-name>
        <conf-name>Proceedings of the 20th ACM SIGKDD International Conference on Knowledge Discovery and Data Mining</conf-name>
        <ext-link ext-link-type="uri" xlink:href="http://www.R-project.org/">http://www.R-project.org/</ext-link>
        <pub-id pub-id-type="doi">10.7554/eLife.42697</pub-id>
        <pub-id pub-id-type="pmid">31038122</pub-id>`,
      'thesis'
    );

    el.setAttribute('id', 'id');
    expect(new Reference(el.querySelector('element-citation'))).toMatchSnapshot();
  });

  it('creates patent reference', () => {
    const el = document.createElement('p');
    el.innerHTML = givenReferenceXml(
      `
        <year iso-8601-date="2015">2015</year>
        <article-title>Imidazopyridine Derivative</article-title>
        <source>World Intellectual Property Organization</source>
        <patent country="Japan">2015087996</patent>
        <ext-link ext-link-type="uri" xlink:href="https://patents.google.com/patent/WO2015087996A1/en">https://patents.google.com/patent/WO2015087996A1/en</ext-link>`,
      'patent'
    );

    el.setAttribute('id', 'id');
    expect(new Reference(el.querySelector('element-citation'))).toMatchSnapshot();
  });

  it('should sort references list', () => {
    const refs = [new Reference(), new Reference(), new Reference(), new Reference()];
    refs[0].type = 'web';
    refs[0].authors = [{ groupName: 'Berk' }];
    (refs[0].referenceInfo as WebReference).year = '2009';
    refs[1].type = 'web';
    refs[1].authors = [{ groupName: 'Twerk' }];
    (refs[1].referenceInfo as WebReference).year = '2001';
    refs[2].type = 'web';
    refs[2].authors = [{ groupName: 'Schwerk' }];
    (refs[2].referenceInfo as WebReference).year = '2011';
    refs[3].type = 'web';
    refs[3].authors = [{ groupName: 'Berk' }];
    (refs[3].referenceInfo as WebReference).year = '2007';

    expect(sortReferencesList(refs)).toEqual([refs[3], refs[0], refs[2], refs[1]]);
  });

  it('should create ref node text', () => {
    const ref = new Reference();
    ref.type = 'web';
    ref.authors = [{ groupName: 'Berk' }];

    expect(ref.getCitationDisplayName()).toBe('Berk');

    set(ref, 'referenceInfo.year', '2009');
    expect(ref.getCitationDisplayName()).toBe('Berk, 2009');

    ref.authors.push({ firstName: 'D', lastName: 'Twerk' });
    expect(ref.getCitationDisplayName()).toBe('Berk and Twerk, 2009');

    ref.authors.push({ firstName: 'B', lastName: 'Schwerk' });
    expect(ref.getCitationDisplayName()).toBe('Berk et al., 2009');
  });

  it('should create reference state from XML', () => {
    const xmlWrapper = document.createElement('div');
    xmlWrapper.innerHTML =
      '<ref id="test">' +
      givenReferenceXml(
        `
      <year iso-8601-date="2015">2015</year>
      <article-title>Imidazopyridine Derivative</article-title>
      <source>World Intellectual Property Organization</source>
      <patent country="Japan">2015087996</patent>
      <ext-link ext-link-type="uri" xlink:href="https://patents.google.com/patent/WO2015087996A1/en">https://patents.google.com/patent/WO2015087996A1/en</ext-link>`,
        'patent'
      ) +
      '</ref>';

    const state = createReferencesState(Array.from(xmlWrapper.querySelectorAll('ref > element-citation')));
    expect(state).toMatchSnapshot();
  });
});

function givenReferenceXml(referenceData: string, type: string): string {
  return `<element-citation publication-type="${type}">
      <person-group person-group-type="author">
          <name>
            <surname>Berk</surname>
            <given-names>V</given-names>
          </name>
          <collab>R Development Core Team</collab>  
      </person-group>
      ${referenceData}
    </element-citation>
  `;
}
