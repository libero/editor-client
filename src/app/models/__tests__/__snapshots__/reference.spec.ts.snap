// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Reference model creates book reference 1`] = `
Reference {
  "_id": "id",
  "_type": "book",
  "authors": Array [
    Object {
      "groupName": "
            Berk
            V
          ",
    },
    Object {
      "groupName": "R Development Core Team",
    },
  ],
  "referenceInfo": Object {
    "chapterTitle": Object {
      "doc": Object {
        "type": "annotatedReferenceInfoDoc",
      },
      "selection": Object {
        "anchor": 0,
        "head": 0,
        "type": "text",
      },
    },
    "doi": "",
    "edition": "4th Edition",
    "editors": Array [],
    "elocationId": "",
    "firstPage": undefined,
    "inPress": false,
    "lastPage": undefined,
    "pmid": "",
    "publisherLocation": "London",
    "publisherName": "Verso",
    "source": Object {
      "doc": Object {
        "type": "annotatedReferenceInfoDoc",
      },
      "selection": Object {
        "anchor": 0,
        "head": 0,
        "type": "text",
      },
    },
    "volume": undefined,
    "year": "2010",
  },
}
`;

exports[`Reference model creates conference reference 1`] = `
Reference {
  "_id": "id",
  "_type": "confproc",
  "authors": Array [
    Object {
      "groupName": "
            Berk
            V
          ",
    },
    Object {
      "groupName": "R Development Core Team",
    },
  ],
  "referenceInfo": Object {
    "articleTitle": Object {
      "doc": Object {
        "content": Array [
          Object {
            "text": "Automated hypothesis generation based on mining scientific literature",
            "type": "text",
          },
        ],
        "type": "annotatedReferenceInfoDoc",
      },
      "selection": Object {
        "anchor": 0,
        "head": 0,
        "type": "text",
      },
    },
    "conferenceDate": "",
    "conferenceLocation": "",
    "conferenceName": Object {
      "doc": Object {
        "content": Array [
          Object {
            "text": "Proceedings of the 20th ACM SIGKDD International Conference on Knowledge Discovery and Data Mining",
            "type": "text",
          },
        ],
        "type": "annotatedReferenceInfoDoc",
      },
      "selection": Object {
        "anchor": 0,
        "head": 0,
        "type": "text",
      },
    },
    "doi": "10.1145/2623330.2623667",
    "elocationId": "",
    "extLink": "",
    "firstPage": "1877",
    "lastPage": "1886",
    "volume": undefined,
    "year": "2014",
  },
}
`;

exports[`Reference model creates data reference 1`] = `
Reference {
  "_id": "id",
  "_type": "data",
  "authors": Array [
    Object {
      "groupName": "
            Berk
            V
          ",
    },
    Object {
      "groupName": "R Development Core Team",
    },
  ],
  "referenceInfo": Object {
    "accessionId": "NM_009324",
    "dataTitle": Object {
      "doc": Object {
        "content": Array [
          Object {
            "text": "Mus musculus T-box 2 (Tbx2), mRNA",
            "type": "text",
          },
        ],
        "type": "annotatedReferenceInfoDoc",
      },
      "selection": Object {
        "anchor": 0,
        "head": 0,
        "type": "text",
      },
    },
    "doi": "",
    "extLink": "http://www.ncbi.nlm.nih.gov/nuccore/120407038",
    "source": Object {
      "doc": Object {
        "type": "annotatedReferenceInfoDoc",
      },
      "selection": Object {
        "anchor": 0,
        "head": 0,
        "type": "text",
      },
    },
    "specificUse": undefined,
    "version": "NM_009324.2",
    "year": "2014",
  },
}
`;

exports[`Reference model creates journal reference 1`] = `
Reference {
  "_id": "id",
  "_type": "journal",
  "authors": Array [
    Object {
      "groupName": "
            Berk
            V
          ",
    },
    Object {
      "groupName": "R Development Core Team",
    },
  ],
  "referenceInfo": Object {
    "articleTitle": Object {
      "doc": Object {
        "content": Array [
          Object {
            "text": "Assembly principles of ",
            "type": "text",
          },
          Object {
            "marks": Array [
              Object {
                "type": "italic",
              },
            ],
            "text": "Vibrio cholerae",
            "type": "text",
          },
          Object {
            "text": " biofilms",
            "type": "text",
          },
        ],
        "type": "annotatedReferenceInfoDoc",
      },
      "selection": Object {
        "anchor": 0,
        "head": 0,
        "type": "text",
      },
    },
    "doi": "",
    "elocationId": "",
    "firstPage": "236",
    "inPress": false,
    "lastPage": "239",
    "pmid": "",
    "source": Object {
      "doc": Object {
        "type": "annotatedReferenceInfoDoc",
      },
      "selection": Object {
        "anchor": 0,
        "head": 0,
        "type": "text",
      },
    },
    "volume": "337",
    "year": "2012",
  },
}
`;

exports[`Reference model creates patent reference 1`] = `
Reference {
  "_id": "id",
  "_type": "patent",
  "authors": Array [
    Object {
      "groupName": "
            Berk
            V
          ",
    },
    Object {
      "groupName": "R Development Core Team",
    },
  ],
  "referenceInfo": Object {
    "articleTitle": Object {
      "doc": Object {
        "content": Array [
          Object {
            "text": "Imidazopyridine Derivative",
            "type": "text",
          },
        ],
        "type": "annotatedReferenceInfoDoc",
      },
      "selection": Object {
        "anchor": 0,
        "head": 0,
        "type": "text",
      },
    },
    "doi": "",
    "extLink": "https://patents.google.com/patent/WO2015087996A1/en",
    "patent": "2015087996",
    "publisherName": "",
    "source": Object {
      "doc": Object {
        "type": "annotatedReferenceInfoDoc",
      },
      "selection": Object {
        "anchor": 0,
        "head": 0,
        "type": "text",
      },
    },
    "year": "2015",
  },
}
`;

exports[`Reference model creates periodical reference 1`] = `
Reference {
  "_id": "id",
  "_type": "periodical",
  "authors": Array [
    Object {
      "groupName": "
            Berk
            V
          ",
    },
    Object {
      "groupName": "R Development Core Team",
    },
  ],
  "referenceInfo": Object {
    "articleTitle": Object {
      "doc": Object {
        "content": Array [
          Object {
            "text": "Obesity affects economic, social status",
            "type": "text",
          },
        ],
        "type": "annotatedReferenceInfoDoc",
      },
      "selection": Object {
        "anchor": 0,
        "head": 0,
        "type": "text",
      },
    },
    "date": "1993-09-09",
    "extLink": "",
    "firstPage": "1",
    "lastPage": "4",
    "source": Object {
      "doc": Object {
        "type": "annotatedReferenceInfoDoc",
      },
      "selection": Object {
        "anchor": 0,
        "head": 0,
        "type": "text",
      },
    },
    "volume": undefined,
  },
}
`;

exports[`Reference model creates preprint reference 1`] = `
Reference {
  "_id": "id",
  "_type": "preprint",
  "authors": Array [
    Object {
      "groupName": "
            Berk
            V
          ",
    },
    Object {
      "groupName": "R Development Core Team",
    },
  ],
  "referenceInfo": Object {
    "articleTitle": Object {
      "doc": Object {
        "content": Array [
          Object {
            "text": "Gender and international diversity improves equity in peer review",
            "type": "text",
          },
        ],
        "type": "annotatedReferenceInfoDoc",
      },
      "selection": Object {
        "anchor": 0,
        "head": 0,
        "type": "text",
      },
    },
    "doi": "10.1101/400515",
    "extLink": "",
    "pmid": "",
    "source": Object {
      "doc": Object {
        "type": "annotatedReferenceInfoDoc",
      },
      "selection": Object {
        "anchor": 0,
        "head": 0,
        "type": "text",
      },
    },
    "year": "2019",
  },
}
`;

exports[`Reference model creates report reference 1`] = `
Reference {
  "_id": "id",
  "_type": "report",
  "authors": Array [
    Object {
      "groupName": "
            Berk
            V
          ",
    },
    Object {
      "groupName": "R Development Core Team",
    },
  ],
  "referenceInfo": Object {
    "doi": "",
    "extLink": "http://www.who.int/malaria/publications/world-malaria-report-2015/en/",
    "isbn": "",
    "pmid": "",
    "publisherLocation": "Geneva",
    "publisherName": "World Health Organization",
    "source": Object {
      "doc": Object {
        "type": "annotatedReferenceInfoDoc",
      },
      "selection": Object {
        "anchor": 0,
        "head": 0,
        "type": "text",
      },
    },
    "volume": undefined,
    "year": "2015",
  },
}
`;

exports[`Reference model creates report reference 2`] = `
Reference {
  "_id": "id",
  "_type": "report",
  "authors": Array [
    Object {
      "groupName": "
            Berk
            V
          ",
    },
    Object {
      "groupName": "R Development Core Team",
    },
  ],
  "referenceInfo": Object {
    "doi": "",
    "extLink": "http://www.who.int/malaria/publications/world-malaria-report-2015/en/",
    "isbn": "",
    "pmid": "",
    "publisherLocation": "Geneva",
    "publisherName": "World Health Organization",
    "source": Object {
      "doc": Object {
        "type": "annotatedReferenceInfoDoc",
      },
      "selection": Object {
        "anchor": 0,
        "head": 0,
        "type": "text",
      },
    },
    "volume": undefined,
    "year": "2015",
  },
}
`;

exports[`Reference model creates software reference 1`] = `
Reference {
  "_id": "id",
  "_type": "software",
  "authors": Array [
    Object {
      "groupName": "
            Berk
            V
          ",
    },
    Object {
      "groupName": "R Development Core Team",
    },
  ],
  "referenceInfo": Object {
    "articleTitle": Object {
      "doc": Object {
        "type": "annotatedReferenceInfoDoc",
      },
      "selection": Object {
        "anchor": 0,
        "head": 0,
        "type": "text",
      },
    },
    "doi": "",
    "extLink": "http://www.R-project.org/",
    "publisherLocation": "Vienna, Austria",
    "publisherName": "R Foundation for Statistical Computing",
    "source": Object {
      "doc": Object {
        "type": "annotatedReferenceInfoDoc",
      },
      "selection": Object {
        "anchor": 0,
        "head": 0,
        "type": "text",
      },
    },
    "version": "",
    "year": "2014",
  },
}
`;

exports[`Reference model creates thesis reference 1`] = `
Reference {
  "_id": "id",
  "_type": "thesis",
  "authors": Array [
    Object {
      "groupName": "
            Berk
            V
          ",
    },
    Object {
      "groupName": "R Development Core Team",
    },
  ],
  "referenceInfo": Object {
    "articleTitle": Object {
      "doc": Object {
        "content": Array [
          Object {
            "text": "Automated hypothesis generation based on mining scientific literature",
            "type": "text",
          },
        ],
        "type": "annotatedReferenceInfoDoc",
      },
      "selection": Object {
        "anchor": 0,
        "head": 0,
        "type": "text",
      },
    },
    "doi": "10.7554/eLife.42697",
    "extLink": "http://www.R-project.org/",
    "pmid": "31038122",
    "publisherLocation": "Vienna, Austria",
    "publisherName": "R Foundation for Statistical Computing",
    "year": "2014",
  },
}
`;

exports[`Reference model creates web reference 1`] = `
Reference {
  "_id": "id",
  "_type": "web",
  "authors": Array [
    Object {
      "groupName": "
            Berk
            V
          ",
    },
    Object {
      "groupName": "R Development Core Team",
    },
  ],
  "referenceInfo": Object {
    "articleTitle": Object {
      "doc": Object {
        "content": Array [
          Object {
            "text": "James Watson had a chance to salvage his reputation on race. He made things worse",
            "type": "text",
          },
        ],
        "type": "annotatedReferenceInfoDoc",
      },
      "selection": Object {
        "anchor": 0,
        "head": 0,
        "type": "text",
      },
    },
    "dateInCitation": "2019-01-01",
    "extLink": "https://www.nytimes.com/2019/01/01/science/watson-dna-genetics-race.html",
    "source": Object {
      "doc": Object {
        "type": "annotatedReferenceInfoDoc",
      },
      "selection": Object {
        "anchor": 0,
        "head": 0,
        "type": "text",
      },
    },
    "year": "2019",
  },
}
`;

exports[`Reference model should create reference state from XML 1`] = `
Array [
  Reference {
    "_id": "test",
    "_type": "patent",
    "authors": Array [
      Object {
        "groupName": "
            Berk
            V
          ",
      },
      Object {
        "groupName": "R Development Core Team",
      },
    ],
    "referenceInfo": Object {
      "articleTitle": Object {
        "doc": Object {
          "content": Array [
            Object {
              "text": "Imidazopyridine Derivative",
              "type": "text",
            },
          ],
          "type": "annotatedReferenceInfoDoc",
        },
        "selection": Object {
          "anchor": 0,
          "head": 0,
          "type": "text",
        },
      },
      "doi": "",
      "extLink": "https://patents.google.com/patent/WO2015087996A1/en",
      "patent": "2015087996",
      "publisherName": "",
      "source": Object {
        "doc": Object {
          "type": "annotatedReferenceInfoDoc",
        },
        "selection": Object {
          "anchor": 0,
          "head": 0,
          "type": "text",
        },
      },
      "year": "2015",
    },
  },
]
`;
